tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+
+		local p_sep = path:sub(-1,-1) == "/"
+		local r_sep = r:sub(1,1) == "/"
+
+		-- avoid duplicated pathseps
+		if p_sep and r_sep then
+			-- remove one of the /
+			return path..r:sub(2)
+		elseif p_sep or r_sep then
+			-- no new / needed
+			return path..r
+		else
+			-- no / yet present
+			return path.."/"..r
+		end
+	end
+end
+
+-----------------
+-- normal code --
+-----------------
+
+-- setup kpse
+kpse.set_program_name("texlua", "memoize-extract.lua")
+
+local find_in
+do
+	local texmf_output_directory = kpse.var_value("TEXMF_OUTPUT_DIRECTORY")
+	local texmfoutput            = kpse.var_value("TEXMFOUTPUT")
+
+	---@param fname string
+	---@return string?
+	---@return string? -- error
+	find_in = function(fname)
+		local abs, err = pathlib.path_is_absolute(fname)
+		if abs == nil then return nil, err end
+		if abs then return fname end
+
+		if texmf_output_directory then
+			local p, err = pathutil.join(texmf_output_directory, fname)
+			if not p then return nil, err end
+			if kpse.in_name_ok_silent_extended(p) then return p end
+		end
+		if not texmf_output_directory then
+			if kpse.in_name_ok_silent_extended(fname) then return fname end
+		end
+		if texmfoutput then
+			local p, err = pathutil.join(texmfoutput, fname)
+			if not p then return nil, err end
+			if kpse.in_name_ok_silent_extended(p) then return p end
+		end
+
+		return fname
+	end
+end
+
+local find_out
+do
+	local texmf_output_directory = kpse.var_value("TEXMF_OUTPUT_DIRECTORY")
+	local texmfoutput            = kpse.var_value("TEXMFOUTPUT")
+
+	---@param fname string
+	---@return string?
+	---@return string?
+	find_out = function(fname)
+		local abs, err = pathlib.path_is_absolute(fname)
+		if abs == nil then return nil, err end
+		if abs then return fname end
+
+		local texmf_od
+		if texmf_output_directory then
+			local p, err = pathutil.join(texmf_output_directory, fname)
+			if not p then return nil, err end
+			texmf_od = p
+			if kpse.out_name_ok_silent_extended(p) then return p end
+		end
+		if not texmf_output_directory then
+			if kpse.out_name_ok_silent_extended(fname) then return fname end
+		end
+		if texmfoutput then
+			local p, err = pathutil.join(texmfoutput, fname)
+			if not p then return nil, err end
+			if kpse.out_name_ok_silent_extended(p) then return p end
+		end
+
+		return texmf_od or fname
+	end
+end
+
+-- setup something like a logging library
+local logging = {
+	file      = nil,
+	header    = "memoize-extract.lua: ",
+	indent    = "",
+	texindent = "",
+}
+do
+	local package_name = "memoize (texlua-based extraction)"
+	local ERROR   = {
+		latex     = function(a) return ("\\PackageError{%s}{%s}{%s}"):format(a.package_name or "", a.short or "", a.long or "") end,
+		plain     = function(a) return ("\\errhelp{%s}\\errmessage{%s: %s}"):format(a.long or "", a.package_name or "", a.short or "") end,
+		context   = function(a) return ("\\errhelp{%s}\\errmessage{%s: %s}"):format(a.long or "", a.package_name or "", a.short or "") end,
```

Comment:
Hmm, this isn't really proper ConTeXt, but ConTeXt also doesn't provide a better way to issue errors from TeX macros, so this is fine I guess.